python/ray/autoscaler/node_provider.py

```diff
@@ -276,3 +278,92 @@ def get_command_runner(self,
             return DockerCommandRunner(docker_config, **common_args)
         else:
             return SSHCommandRunner(**common_args)
+
+class NodeProviderCache:
+    """Base class for storing cached node information, used to avoid excessive
+    API calls to cloud providers in some cases.
+
+    This implementation stores all information in memory. 
+    """
+
+    def __init__(self):
+        # tag_map is supposed to be the superset of node_map.
+        self.node_map = {}
+        self.tag_map = {}
+        self.lock = threading.Lock()
+
+    def get_node(self, node_id):
+        with lock:
+            return self.node_map.get(node_id)
+
+    def set_node(self, node_id, node):
+        with lock:
+            self.node_map[node_id] = node
+            self.tag_map.setdefault(node_id, {})
+
+    def node_exists(self, node_id):
+        with lock:
+            return node_id in self.node_map
+
+    def get_tags(self, node_id):
+        with lock:
+            return self.tag_map.get(node_id)
+
+    def set_tags(self, node_id, tags):
+        with lock:
+            self.tag_map[node_id] = tags
+
+    def update_tags(self, node_id, tags):
+        with lock:
+            self.tag_map[node_id].update(tags)
+
+    def tags_exists(self, node_id):
+        with lock:
+            return node_id in self.tags_map
+
+    def get_nodes_by_tags(self, tag_filters=None):
+        result = {}
+        with lock:
+            for node_id, tags in self.tag_map.items():
+                tag_match = True
+                for k, v in tag_filters.items():
+                    if tags.get(k) != v:
+                        tag_match = False
+                        break
+                if tag_match:
+                    if node_id in self.node_map:
+                        result[node_id] = copy.deepcopy(node)
+        
+        return result
+
+    def detele_node_and_tags(self, node_id):
+        with lock:
+            if node_id in self.node_map:
+                del self.node_map[node_id]
+            if node_id in self.tag_map:
+                del self.tag_map[node_id]
+
+    def cleanup_by_tags(self, tag_filters=None):
```

Comment:
What is this for, we didn't seem to need it before.

Reply:
Current behavior is, whenever `non_terminated_nodes()` is called, we remove everything in cache, and depends on tag_filter parameter, sometimes we only repopulate part of the cache (head node or worker nodes).
I want to change that to: we only erase the cache for the type of nodes that will be repopulated.

Reply:
I see. To avoid this PR be both a refactoring and high-risk change, please split this optimization out into a follow-up PR.

In the followup, we should also make sure to use the tag cache in the MockProvider class in order for it to be exercised by the unit tests.